pkg/schemacompat/schemacompat_test.go

```diff
@@ -186,6 +190,244 @@ func TestCompatibility(t *testing.T) {
+				"existing": {Type: "boolean"},
+			},
+		},
+		new: &apiextensionsv1.JSONSchemaProps{
+			Type: "object",
+			Properties: map[string]apiextensionsv1.JSONSchemaProps{
+				"existing": {Type: "boolean"},
+				"new":      {Type: "number"},
+			},
+		},
+		// LCD is the same as existing.
+		wantLCD: &apiextensionsv1.JSONSchemaProps{
+			Type: "object",
+			Properties: map[string]apiextensionsv1.JSONSchemaProps{
+				"existing": {Type: "boolean"},
+			},
+		},
+	}, {
+		desc: "new has more properties, existing contains array",
+		existing: &apiextensionsv1.JSONSchemaProps{
+			Type: "array",
+			Items: &apiextensionsv1.JSONSchemaPropsOrArray{
+				Schema: &apiextensionsv1.JSONSchemaProps{
+					Type: "object",
+					Properties: map[string]apiextensionsv1.JSONSchemaProps{
+						"existing": {Type: "integer"},
+					},
+				},
+			},
+		},
+		new: &apiextensionsv1.JSONSchemaProps{
+			Type: "array",
+			Items: &apiextensionsv1.JSONSchemaPropsOrArray{
+				Schema: &apiextensionsv1.JSONSchemaProps{
+					Type: "object",
+					Properties: map[string]apiextensionsv1.JSONSchemaProps{
+						"existing": {Type: "integer"},
+						"new":      {Type: "number"},
+					},
+				},
+			},
+		},
+		// LCD is the same as existing.
+		wantLCD: &apiextensionsv1.JSONSchemaProps{
+			Type: "array",
+			Items: &apiextensionsv1.JSONSchemaPropsOrArray{
+				Schema: &apiextensionsv1.JSONSchemaProps{
+					Type: "object",
+					Properties: map[string]apiextensionsv1.JSONSchemaProps{
+						"existing": {Type: "integer"},
+					},
+				},
+			},
+		},
+	}, {
+		desc: "new has more properties, existing contains additional properties",
+		existing: &apiextensionsv1.JSONSchemaProps{
+			Type: "object",
+			AdditionalProperties: &apiextensionsv1.JSONSchemaPropsOrBool{
+				Schema: &apiextensionsv1.JSONSchemaProps{
+					Type: "object",
+					Properties: map[string]apiextensionsv1.JSONSchemaProps{
+						"subProp1": {Type: "string"},
+					},
+				},
+			},
+		},
+		new: &apiextensionsv1.JSONSchemaProps{
+			Type: "object",
+			AdditionalProperties: &apiextensionsv1.JSONSchemaPropsOrBool{
+				Schema: &apiextensionsv1.JSONSchemaProps{
+					Type: "object",
+					Properties: map[string]apiextensionsv1.JSONSchemaProps{
+						"subProp1": {Type: "string"},
+						"subProp2": {Type: "string"},
+					},
+				},
+			},
+		},
+		// LCD is the same as existing.
+		wantLCD: &apiextensionsv1.JSONSchemaProps{
+			Type: "object",
+			AdditionalProperties: &apiextensionsv1.JSONSchemaPropsOrBool{
+				Allows: true,
+				Schema: &apiextensionsv1.JSONSchemaProps{
+					Type: "object",
+					Properties: map[string]apiextensionsv1.JSONSchemaProps{
+						"subProp1": {Type: "string"},
+					},
+				},
+			},
+		},
+	}, {
+		desc: "new has more properties, existing contains additional properties boolean",
+		existing: &apiextensionsv1.JSONSchemaProps{
+			Type: "object",
+			AdditionalProperties: &apiextensionsv1.JSONSchemaPropsOrBool{
+				Allows: true,
```

Comment:
do we allow that? I don't we do. That would mean polymorphic string maps, i.e. `map[string]interface{}`.

Reply:
It is allowed in here https://github.com/kcp-dev/kcp/blob/f1145a6ec820b66420d651c0bce83a3ccf30769a/pkg/schemacompat/schemacompat.go#L372.

While adding unit tests, I assumed that there is no bug. What do you think?

Reply:
Note we have a number of restrictions that are not visible in the types, but are done in the API validation phase. The structural schema types are much more restricted in this regard.